src/review-todo.js

```diff
@@ -0,0 +1,212 @@
+let isDate = require("date-fns/isDate");
+
+function todo(todoName = "New Todo") {
+	let title = todoName,
+		description = "",
+		urgency = 0,
+		isDone = false,
+		dueDate = todoDate(),
+		scheduleDate = todoDate(),
+		tags = tagList();
+
+	let todoObj = Object.create(todo.proto);
+	Object.assign(todoObj, {
+		title,
+		description,
+		isDone,
+		urgency,
+		dueDate,
+		scheduleDate,
+		tags,
+	});
+
+	return todoObj;
+}
+
+todo.proto = {};
+
+function project(projectName = "New Project") {
+	let proto = Object.assign(project.proto, { isChild });
+	let projectObj = Object.create(proto);
+
+	let title = projectName,
+		description = "",
+		favorite = false,
+		urgency = 0,
+		isDone = false,
+		dueDate = todoDate(),
+		scheduleDate = todoDate(),
+		tags = tagList(),
+		_todoList = createTodoList(projectObj),
+		todoList = Object.create(_todoList),
+		createTag = tagGenerator("project").bind(projectObj);
+
+	todoList.add = (proj) => {
+		if (
+			Object.getPrototypeOf(proj) === project.proto &&
+			projectObj.isChild(proj)
+		) {
+			throw "Looping projects";
+		}
+		_todoList.add(proj);
+	};
+
+	function isChild(project) {
+		let projects = this.tags
+			.list()
+			.filter((tag) => tag.tagType === "project");
+		if (!projects) return false;
+		if (projects.some((proj) => proj.id === project)) return true;
+		if (projects.some((proj) => proj.id.isChild(project))) return true;
```

Comment:
Can we use logical operators to remove some of the if statements here? Maybe
```
function isChild(project) {
  return  !!projects && validRelation(projects, project)
}

function validRelation(projects, project) {
  return projects.some((proj) => proj.id === project)) || projects.some((proj) => proj.id.isChild(project)) 
} 
```

Basically, make complex conditional logic hide behind well-named functions and return your expression result instead of explicitly returning true or false.

Reply:
Do you happen to know if there is a naming convention for functions like these? Should I be using `childObject.isChild(parentObject)` or `parentObject.isChild(childObject)`?

Reply:
The important convention is using `is` or `has` for boolean methods. The actual method name should be whatever you can use to make it unambiguous. `childObject.isChildOf(parentObject)`, `childObject.hasParent(parentObject)`, or `parentObject.hasChild(childObject)` all seem clearer than `isChild` to me. Naming things well is one of the hard problems in this work, though, and it's normal to go back and forth a bit while honing on the reasonable thing.